lowcar/devices/KoalaBear/PID.cpp

```diff
@@ -1,125 +1,61 @@
-
-
-	this->prevPos = this->enc->read();
-	//prevVel = setPoint;
-	if (out > 1)
-	{
-		return 1;
-	}
-	else if (out < -1)
-	{
-		return -1;
-	}
-	return out;
-}
-
-void PID::setCoefficients(float KP, float KI, float KD) {
-	this->kp = KP;
-	this->ki = KI;
-	this->kd = KD;
-}
-
-void PID::setSetpoint(float sp) { this->setPoint = sp; }
-
-float PID::getKP() { return this->kp; }
-float PID::getKI() { return this->ki; }
-float PID::getKD() { return this->kd; }
-
-float PID::readPos() { return this->enc->read(); }
-float PID::readVel() { return 0.00; }
-
-/****************************************************
-* 				ENCODER FUNCTIONS
-****************************************************/
-long res = 100;
-long interval_us = res*(2500/11); //interval in us; 2500/11 == 1000000 (1 sec) / 4400 (tick range)
-
-void PID::encoderSetup() {
-  pinMode(encoder0PinA,INPUT);
-  pinMode(encoder0PinB,INPUT);
-}
-
-void PID::resetEncoder() {
-  enc->write(0);
-}
-
-// Updates the encoder's velocity by calculating the position slope.
-void PID::updateVel() {
-  float enc_reading = this->enc->read();
-  this->prevVel = ((enc_reading - this->prevPos)*1000000)/((float) interval_us);
-  this->prevPos = enc_reading; // Save the current pos
-}
-
-/**************
-Original Encoder
-**************/
-/*
-long res = 100;
-long interval_us = res*(2500/11); //interval in us; 2500/11 == 1000000 (1 sec) / 4400 (tick range)
-float pos = 0;
-float vel = 0;
-volatile signed long old_encoder0Pos = 0;
-
-void PID::updateEncoder() {
-  updatePos();
-  updateVel();
-}
-
-void PID::encoderSetup() {
-  pinMode(encoder0PinA,INPUT);
-  pinMode(encoder0PinB,INPUT);
-}
-
-void PID::resetEncoder() {
-  enc->write(0);
-}
-
-float PID::readPos() {
-  return pos;
-}
-
-float PID::readVel() {
-  return vel;
-}
-
-void PID::updatePos() {
-  pos = enc->read();
+ * Arguments:
+ *    curr_pos: current value of the encoder, as a float
+ */
+float PID::compute(float curr_pos) {
+    unsigned long curr_time = micros(); // get the current time
+    float interval_secs = ((float)(curr_time - this->prev_time)) / 1000000.0; // compute time passed between loops, in seconds
+    float desired_pos = prev_desired_pos + (duty_cycle_to_tps(this->target_speed) * interval_secs); // compute the desired position at this time
+    float error = desired_pos - curr_pos; // compute the error as the set point (desired position) - process variable (current position)
+    this->integral += error * (interval_secs); //compute the new value of this->integral using right-rectangle approximation
+    
+    // store the current values into previous values for use on next loop
+    this->prev_error = error;
+    this->prev_pos = curr_pos;
+    this->prev_desired_pos = desired_pos;
+    this->prev_time = curr_time;
+    
+    // return kp * error + ki * integral of error * kd * "derivative" of error
+    return (this->kp * error) + (this->ki * this->integral) + (this->kd * (error - this->prev_error));
```

Comment:
you need to divide the derivative term by `interval_secs`. you also are using the new `prev_error`, which is just equal to `error` so that won't work.

Reply:
Ah, you're right. Lemme fix this